### DISCUSSION TITLE
git vs github for remote repositories and branches

### DISCUSSION THREAD

**Discussion by daldr-ntml** [community/community#64867] [New to GitHub] [answered]
Hi, I am relatively new to git and github. I am looking at using git as a replacement for Subversion, for small team development.

We would certainly need to use a remote repository, to mimic Subversion's centralised repository.  I think a simple workflow would be sufficient, such as the GitHub flow.

If we worked purely in git I think we would have to create remote branches to connect to the remote repository. That, together with the complexities of merging and rebasing, looks quite daunting.

Does GitHub actually help to make this process easier?  I understand that GitHub hosts repositories but I don't really understand how it can replace native git commands and improve the user experience.

**Comment by Jaber-Saed** [comment, marked as answer]
Absolutely, using Git and GitHub for small team development can greatly improve your collaboration and version control processes. It's understandable that transitioning from Subversion to Git might seem a bit overwhelming at first, but Git and platforms like GitHub offer a more flexible and powerful approach to version control.

Here's how GitHub can make the process easier and improve your experience:

Remote Repository Hosting: GitHub provides a remote repository where your team can collaborate on code. This central repository serves as a single source of truth, similar to how Subversion's central repository works.

Simplified Workflow: GitHub simplifies and streamlines the process of collaborating with others using Git. While the basic Git workflow involves branches and merges, GitHub's user interface makes these actions more intuitive and visually accessible.

Branching Strategies: GitHub makes it easier to create and manage branches. The GitHub flow, which is a simplified Git branching strategy, encourages using branches for features, bug fixes, and experiments. GitHub's interface simplifies branch creation, making it easier to manage parallel development.

Pull Requests: One of the key features of GitHub is the pull request (PR) mechanism. Pull requests facilitate code review and collaboration. Instead of directly merging changes into the main branch, you propose changes via a pull request, allowing for discussion, feedback, and ensuring quality before merging.

Code Review: GitHub's pull request system includes features for code review, allowing team members to provide feedback, ask questions, and suggest changes directly on the code. This ensures that code quality is maintained and helps catch errors early.

Automatic Merging: GitHub simplifies merging by providing options for automatic merging when changes can be merged without conflicts. If conflicts do arise, GitHub provides a visual interface to resolve them.

Visual Diffing: GitHub's user interface includes visual diffing tools that make it easier to review changes between different branches, commits, or pull requests. This helps you understand the impact of your changes and any conflicts that might arise.

Integration with Third-Party Tools: GitHub integrates well with various tools, such as continuous integration (CI) systems and project management tools, enhancing your development workflow.

While GitHub simplifies many aspects of Git, it's important to note that learning the basics of Git is still essential for using GitHub effectively. GitHub enhances Git by providing a user-friendly interface for collaboration, code review, and project management, but understanding Git's underlying concepts is beneficial for handling more complex scenarios.

Overall, GitHub helps streamline the Git workflow, making it more approachable and user-friendly, especially for small teams collaborating on code. It's worth exploring GitHub's features and gradually incorporating them into your team's development process.

**Reply by daldr-ntml** [reply]
Thank you for your answer.

**Comment by spenserblack** [comment]
FYI you don't really need to rebase, and I wouldn't really recommend it for git beginners. It's a powerful tool to keep your git history clean, but even a messy git history is useful :laughing:

GitHub certainly makes Git more beginner-friendly. The UI elements will steer you towards the most commonly used git features, and all you really need to track your history. GitHub also functions as an issue tracker, so you can raise an issue, discuss it, make a PR linking the issue, discuss the code changes in a reivew, and merge when ready. GitHub also provides GitHub Actions so you can run automated tasks -- probably the most common would be running build and test commands on PRs to assert that the PR doesn't break anything. Though my personal favorite is GitHub Codespaces: a cloud computing development instance with standardized configurations. This way you know you have the correct tools installed when working in a codespace.

I'd recommend looking into forking instead of working with a centralized repo, as it's going to be a lot more mistake-proof -- if someone messes up the repo severely, they've only messed up *their fork,* not the main repo. There really isn't much of a difference in your workflow if everyone works on their personal forks vs. everyone working on their own branches in the same repo. Code, commit, push, make a PR, merge the PR. Forks will give your contributors more freedom to play around. For example they can experiment with force-pushing without worrying about destroying the main repo's history.

**Reply by daldr-ntml** [reply]
Thanks for your answer.

**Comment by noxiouswire** [comment]
It's help me alot. Thx

<a href="https://blobopera.io">blob opera</a>
